File: 29Items-Projects/1Java/4-Document-Approval-Workflow/backend/src/main/java/com/approval/workflow/handlers/ErrorHandler.java
package com.approval.workflow.handlers;

import io.vertx.core.Handler;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.RoutingContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;

/**
 * Centralized failure handler formatting RFC 7807 problem details.
 */
public class ErrorHandler implements Handler<RoutingContext> {
    private static final Logger log = LoggerFactory.getLogger(ErrorHandler.class);

    @Override
    public void handle(RoutingContext ctx) {
        Throwable failure = ctx.failure();
        int statusCode = ctx.statusCode() > 0 ? ctx.statusCode() : 500;
        String message = failure != null ? failure.getMessage() : "Internal Server Error";

        if (statusCode == 500 && failure != null) {
            log.error("Unhandled server exception at {}: {}", ctx.request().path(), failure.getMessage(), failure);
        } else {
            log.warn("Handled client error at {} [status={}]: {}", ctx.request().path(), statusCode, message);
        }

        JsonObject errorResponse = new JsonObject()
                .put("status", statusCode)
                .put("error", getErrorType(statusCode))
                .put("message", message)
                .put("path", ctx.request().path())
                .put("timestamp", Instant.now().toString());

        ctx.response()
                .setStatusCode(statusCode)
                .putHeader("Content-Type", "application/json")
                .end(errorResponse.encode());
    }

    private String getErrorType(int statusCode) {
        return switch (statusCode) {
            case 400 -> "Bad Request";
            case 401 -> "Unauthorized";
            case 403 -> "Forbidden";
            case 404 -> "Not Found";
            case 409 -> "Conflict";
            case 422 -> "Unprocessable Entity";
            default -> "Internal Server Error";
        };
    }
}

File: 29Items-Projects/9Web-design/1Design-System-Component-Library/src/api/error-handler.ts
import type { ErrorRequestHandler } from 'express';
import { isDomainError } from '../services/domain-errors';

export const errorHandler: ErrorRequestHandler = (error, _request, response, next) => {
  void next;

  if (isDomainError(error)) {
    response.status(error.statusCode).json({
      error: {
        code: error.code,
        message: error.message,
        details: error.details
      }
    });
    return;
  }

  response.status(500).json({
    error: {
      code: 'INTERNAL_SERVER_ERROR',
      message: process.env.NODE_ENV === 'production' ? 'Unexpected server error' : error.message
    }
  });
};
